### PR TITLE
✨ Article catalog and tag 

### PR DESCRIPTION
**Pull request by shirne**
- [x] search input (interface)
- [x] catalog list
- [x] catalog switch
- [x] sort & search header flow
- [x] SortType & localizations

search would work in a new PR

|Tab Switch|Header show&hide with scroll|
|:-:|:-:|
|<video src="https://user-images.githubusercontent.com/2263157/186442627-2162940b-0d01-401e-a8f6-91bca4be2b5f.mp4" />|<video src="https://user-images.githubusercontent.com/2263157/186545209-a5b2654c-0b60-4917-86ac-1aa14dec44eb.mp4" />|

### CODE REVIEW

File: lib/apis/recommend_api.dart [hunk 1]
```diff
@@ -14,6 +14,45 @@ class RecommendAPI {
   static const String _articles = '$_api/article';
   static const String _pins = '$_api/short_msg';
 
+  static Future<ResponseModel<FeedModel>> getArticles({
+    bool isFollow = false,
+    String? lastId,
+    String? cateId,
```
Comment (AlexV525) [inline lib/apis/recommend_api.dart]: Should be `categoryId`

File: lib/apis/recommend_api.dart [hunk 2]
```diff
@@ -14,6 +14,45 @@ class RecommendAPI {
...
+    String? tagId,
+    int sortType = 200,
+    int limit = 20,
+    int idType = 2,
```
Comment (AlexV525) [inline lib/apis/recommend_api.dart]: `idType` doesn't indicate valid things. Make it self-describe or add comments for it.

File: lib/apis/recommend_api.dart [hunk 3]
```diff
@@ -29,6 +68,83 @@ class RecommendAPI {
...
+    String? lastId,
+    required String cateId,
+    required String tagId,
+    int idType = 2,
+    int limit = 20,
+    int sortType = 200,
+  }) {
+    return HttpUtil.fetchModel(
+      FetchType.post,
+      url: '$_articles/recommend_cate_tag_feed',
```
Comment (AlexV525) [inline lib/apis/recommend_api.dart]: Url can be combined using `StringBuffer` with conditions.

File: lib/apis/recommend_api.dart (hunk 1 above)
Comment (shirne) [reply lib/apis/recommend_api.dart]: api argument is  cate_id

**Comment by shirne** [comment]
#22

File: lib/apis/recommend_api.dart (hunk 1 above)
Comment (AlexV525) [reply lib/apis/recommend_api.dart]: We don't need to follow the previous naming if we have a better one.

File: lib/models/loading_base.dart [hunk 4]
```diff
@@ -62,9 +62,9 @@ class LoadingBase<T extends DataModel> extends LoadingMoreBase<T> {
         addAll(response.models!);
         total = response.total!;
         page = response.page ?? newPage;
-        lastId = cursorType == CursorType.json
-            ? (_decodeCursor(response.cursor)?['v'])
-            : response.cursor;
+        lastId = getCursorType?.call() != CursorType.json
```
Comment (AlexV525) [inline lib/models/loading_base.dart]: Use straight conditions as much as possible.

```suggestion
        lastId = getCursorType?.call() == CursorType.json
```

File: lib/pages/home/articles.dart [hunk 5]
```diff
@@ -102,45 +109,56 @@ class _ArticlesPageState extends State<ArticlesPage>
...
-        }
-      },
-      decoration: InputDecoration(
-        isDense: true,
-        prefixIcon: Icon(
-          Icons.search,
+    return Center(
+      child: TextField(
+        controller: searchTextController,
+        style: context.textTheme.subtitle2?.copyWith(
```
Comment (AlexV525) [inline lib/pages/home/articles.dart]: `subtitle2` will be soon deprecated. Use `titleSmall` instead.

File: lib/models/loading_base.dart (hunk 4 above)
Comment (shirne) [reply lib/models/loading_base.dart]: ```(getCursorType?.call() ?? CursorType.json ) == CursorType.json```
Maybe like this ?

**Comment by shirne** [comment]
Tags dropdown and fold up effect
<video src="https://user-images.githubusercontent.com/2263157/186822104-a2a8cebf-e43f-4838-8dbe-64c6615479d6.mp4" />

**Comment by shirne** [comment]
https://user-images.githubusercontent.com/2263157/187703436-c8c61340-0422-4476-892a-e28c854f2a68.mp4

**Comment by shirne** [comment]
- hide search header when scrolling instead of receive scroll notification

https://user-images.githubusercontent.com/2263157/188448400-aec88224-c322-4631-b0c7-2f01e6aaac8a.mp4

**Comment by shirne** [comment]
https://user-images.githubusercontent.com/2263157/189116992-4253cc94-e24b-4a2e-8adc-d7eb8ec9c2ad.mp4

**Review by Passer-by — APPROVED** [review]

**Pull request merged**